components/client/src/main/java/com/hotels/styx/client/Transport.java

```diff
@@ -15,76 +15,41 @@
  */
 package com.hotels.styx.client;
 
-import com.hotels.styx.api.Id;
 import com.hotels.styx.api.LiveHttpRequest;
 import com.hotels.styx.api.LiveHttpResponse;
 import com.hotels.styx.api.ResponseEventListener;
-import com.hotels.styx.api.exceptions.NoAvailableHostsException;
 import com.hotels.styx.client.connectionpool.ConnectionPool;
 import rx.Observable;
 
-import java.util.Optional;
-
-import static java.util.Objects.requireNonNull;
-
 /**
  * Encapsulates a single connection to remote server which we can use to send the messages.
  */
 class Transport {
-    private final Id appId;
-    private final CharSequence originIdHeaderName;
-
-    public Transport(Id appId, CharSequence originIdHeaderName) {
-        this.appId = requireNonNull(appId);
-        this.originIdHeaderName = requireNonNull(originIdHeaderName);
-    }
-
-    public HttpTransaction send(LiveHttpRequest request, Optional<ConnectionPool> origin, Id originId) {
-        Observable<Connection> connection = connection(request, origin);
+    public HttpTransaction send(LiveHttpRequest request, ConnectionPool connectionPool) {
+        Observable<Connection> connection = connectionPool.borrowConnection();
 
         return new HttpTransaction() {
             @Override
             public Observable<LiveHttpResponse> response() {
                 return connection.flatMap(connection -> {
-                    Observable<LiveHttpResponse> responseObservable = connection.write(request)
-                            .map(response -> addOriginId(originId, response));
+                    Observable<LiveHttpResponse> responseObservable = connection.write(request);
 
                     return ResponseEventListener.from(responseObservable)
-                            .whenCancelled(() -> closeIfConnected(origin, connection))
-                            .whenResponseError(cause -> closeIfConnected(origin, connection))
-                            .whenContentError(cause -> closeIfConnected(origin, connection))
-                            .whenCompleted(() -> returnIfConnected(origin, connection))
+                            .whenCancelled(() -> closeIfConnected(connectionPool, connection))
+                            .whenResponseError(cause -> closeIfConnected(connectionPool, connection))
+                            .whenContentError(cause -> closeIfConnected(connectionPool, connection))
+                            .whenCompleted(() -> returnIfConnected(connectionPool, connection))
                             .apply();
                 });
             }
 
-            private synchronized void closeIfConnected(Optional<ConnectionPool> connectionPool, Connection connection) {
-                if (connection != null && connectionPool.isPresent()) {
```

Comment:
Have we confirmed that the connections are never null? It would make sense, but thought it worth checking.

Reply:
A `null` value wouldn't be emitted under any circumstances (that would be a bug elsewhere in styx). 

Note that the Reactive Streams specification specifically forbids `null` values as events (Rule 2.13 https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.2/README.md). Although we are still in RxJava 1.0, our intention is to fully migrate to reactive-streams compatible implementation (Flux), ensuring it cannot be null.